cuda/matrix/ell_kernels.cu

```diff
@@ -151,17 +270,98 @@ void advanced_spmv(std::shared_ptr<const CudaExecutor> exec,
+            a->get_size()[0], as_cuda_type(alpha->get_const_values()),
+            as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(beta->get_const_values()),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 4:
+        abstract_spmv<4><<<grid_size, block_size, 0, 0>>>(
+            a->get_size()[0], as_cuda_type(alpha->get_const_values()),
+            as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(beta->get_const_values()),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 8:
+        abstract_spmv<8><<<grid_size, block_size, 0, 0>>>(
+            a->get_size()[0], as_cuda_type(alpha->get_const_values()),
+            as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(beta->get_const_values()),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 16:
+        abstract_spmv<16><<<grid_size, block_size, 0, 0>>>(
+            a->get_size()[0], as_cuda_type(alpha->get_const_values()),
+            as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(beta->get_const_values()),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 32:
+        if (atomic) {
+            dense::scale(exec, beta, c);
+            abstract_spmv<32, true><<<grid_size, block_size, 0, 0>>>(
+                a->get_size()[0], as_cuda_type(alpha->get_const_values()),
+                as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+                a->get_stride(), a->get_num_stored_elements_per_row(),
+                as_cuda_type(b->get_const_values()), b->get_stride(),
+                as_cuda_type(beta->get_const_values()),
+                as_cuda_type(c->get_values()), c->get_stride());
+        } else {
+            abstract_spmv<32><<<grid_size, block_size, 0, 0>>>(
+                a->get_size()[0], as_cuda_type(alpha->get_const_values()),
+                as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+                a->get_stride(), a->get_num_stored_elements_per_row(),
+                as_cuda_type(b->get_const_values()), b->get_stride(),
+                as_cuda_type(beta->get_const_values()),
+                as_cuda_type(c->get_values()), c->get_stride());
+        }
+        break;
```

Comment:
You do this huge switch-case two times. Additionally, it is very redundant. I would prefer a version where you use something similar to `cuda/preconditioner/jacobi_kernels.cu` with the `compiled_kernels`.

Goran used `GKO_ENABLE_IMPLEMENTATION_SELECTION` from `core/synthesizer/implementation_selection.hpp`, which even has a case where nothing matches the given list (which is also why I would prefer a solution similar to that).
The mapping is done at run time (what you need), but the list (against the matching is done) must be defined at compile time (also perfect in this case).